### PR TITLE
apis and tests for ldap

### PR DESCRIPTION
**Pull request by waghswapnil**
- apis and tests for ldap
- new apis/utils.py for utilities used by all apis
- skip setting None values in ManagedObject.__set_prop (this needs to be redesigned)

Signed-off-by: Swapnil Wagh <waghswapnil@gmail.com>

### CODE REVIEW

File: imcsdk/apis/admin/ldap.py [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+
+def _set_ldap_servers(mo, ldap_servers):
+    if len(ldap_servers) > len(_LDAP_SERVERS):
+        raise ImcOperationError("Configure LDAP", "Cannot configure more than"
+                                "%d servers" % len(_LDAP_SERVERS))
+    servers = {}
+    ports = {}
+    for server in ldap_servers:
+        servers['ldap_server' + str(server['id'])] = server['ip']
+        ports['ldap_server_port' + str(server['id'])] = str(server['port'])
```
Comment (vvb) [inline imcsdk/apis/admin/ldap.py:43]: The below compresses your 5 lines into 2
servers = {'ldap_server'+str(x['id']): x['ip'] for x in ldap_servers}
ports = {'ldap_server_port'+str(x['id']): x['port'] for x in ldap_servers}

File: imcsdk/apis/admin/ldap.py [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+        }
+
+    mo.set_prop_multiple(**params)
+
+    if ldap_servers:
+        _set_ldap_servers(mo, ldap_servers)
+
+    mo.set_prop_multiple(**kwargs)
+    handle.set_mo(mo)
+    return handle.query_dn(mo.dn)
```
Comment (vvb) [inline imcsdk/apis/admin/ldap.py:136]: neat!

File: imcsdk/apis/admin/ldap.py [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+
+    if _is_valid_arg('group_nested_search', kwargs):
+        params['group_nested_search'] = str(kwargs.pop('group_nested_search'))
+
+    if not mo.check_prop_match(**params):
+        return False
+
+    # pop the password property if it exists
+    if _is_valid_arg('password', kwargs):
+        kwargs.pop('password')
```
Comment (vvb) [inline imcsdk/apis/admin/ldap.py:193]: any specific reason for using `kwargs.pop`. The only scenario I have found it useful is to defect spurious input. But here we aren't checking anything. Given that we use `kwargs.get` everywhere else, should we just stick to it. your call.

Comment (vvb) [reply imcsdk/apis/admin/ldap.py:193]: Guess, I spoke to soon in the code review process. you next function sort of makes it clear, why you need it. :-)

File: imcsdk/apis/admin/ldap.py [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+    """
+
+    mo = _get_mo(handle, dn=LDAP_DN)
+    return mo.admin_state.lower() == "enabled"
+
+
+def _get_free_ldap_role_group_id(handle):
+    mos = handle.query_classid('AaaLdapRoleGroup')
+    for mo in mos:
+        if not mo.name and not mo.domain:
```
Comment (vvb) [inline imcsdk/apis/admin/ldap.py:256]: `not a and not b` -> `not(a or b)`

File: imcsdk/apis/admin/ldap.py [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+
+    Returns:
+        bool
+    """
+    mo = _get_mo(handle, dn="sys/ldap-ext/ldap-ca-cert-mgmt")
+    return mo.binding_certificate.lower() == "enabled"
+
+
+def ldap_certificate_download(handle, user=None, pwd=None, remote_server=None,
+                              remote_file=None, protocol='tftp', **kwargs):
```
Comment (vvb) [inline imcsdk/apis/admin/ldap.py]: remote_server, remote_file is always required. right? If so, it should not have `None` as default. user/pwd may not be required for `http` case i guess.

File: imcsdk/apis/admin/ldap.py [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+    }
+    mo.set_prop_multiple(**params)
+    mo.set_prop_multiple(**kwargs)
+    handle.set_mo(mo)
+    return handle.query_dn(mo.dn)
+
+
+def ldap_certificate_export(handle, user=None, pwd=None, remote_server=None,
+                            remote_file=None, protocol='tftp', **kwargs):
+    """
```
Comment (vvb) [inline imcsdk/apis/admin/ldap.py:467]: same as above

File: tests/server/test_ldap.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    match, mo = ldap_role_group_exists(handle, domain='abcd.pqrs.com', name='abcd', role='user')
+    assert_equal(match, False)
+
+
+def test_ldap_cert_mgmt_enable():
+    ldap_certificate_management_enable(handle)
+    assert_equal(is_ldap_certificate_management_enabled(handle), True)
+
+
+def test_ldap_cert_mgmt_disable():
```
Comment (vvb) [inline tests/server/test_ldap.py:122]: as a part of the tests, can you also write some error cases. input missing etc..

**Review by vvb — CHANGES_REQUESTED** [review]
@waghswapnil thanks for the PR. some minor comments.

File: imcsdk/apis/admin/ldap.py (hunk 1 above)
Comment (waghswapnil) [reply imcsdk/apis/admin/ldap.py:43]: right, but wanted to avoid 2 loops

File: imcsdk/apis/admin/ldap.py (hunk 4 above)
Comment (waghswapnil) [reply imcsdk/apis/admin/ldap.py:256]: right, but i always find the former more readable

File: imcsdk/apis/admin/ldap.py (hunk 5 above)
Comment (waghswapnil) [reply imcsdk/apis/admin/ldap.py]: will make this change

**Review by vvb — APPROVED** [review]
LGTM

**Pull request merged**